docs/resources/server/enchantments/index.md

```diff
@@ -0,0 +1,307 @@
+  // In vanilla, these are: `any`, `hand`, `mainhand`, `offhand`, `armor`, `feet`, `legs`, `chest`, `head`, and  `body`.
+  "slots": [
+    "mainhand"
+  ],
+
+  // The effects that this enchantment provides as a map of enchantment effect components (read on).
+  "effects": {
+    "examplemod:custom_effect": [
+      {
+        "effect": {
+          "type": "minecraft:add",
+          "value": {
+            "type": "minecraft:linear",
+            "base": 1,
+            "per_level_above_first": 1
+          }
+        }
+      }
+    ]
+  }
+}
+```
+
+### Enchantment Costs
+
+The `max_cost` and `min_cost` fields specify boundaries for how much enchanting power is needed to create this enchantment. There is a somewhat convoluted procedure to actually make use of these values, however.
+
+First, the table takes into account the return value of `IBlockStateExtension#getEnchantPowerBonus()` for the surrounding blocks. From this, it derives a 'base cost' for each slot. This cost is shown in-game as the green numbers besides the enchantments in the menu. For each enchantment, the base cost is modified twice by a random value derived from the item's enchantability (its return value from `IItemStackExtension#getEnchantmentValue()`), like so:
+
+`(Modified Cost) = (Base Cost) + random.nextInt(e / 4 + 1) + random.nextInt(e / 4 + 1)`, where `e` is the enchantability score.
+
+This modified cost is adjusted up or down by a random 15%, and then is finally used to choose an enchantment. 
+
+In practical terms, this means that the cost values in your enchantment definition might be above 30, sometimes far above. For example, with an enchantability 10 item, the table could produce enchantments up to 1.15 * (30 + 2 * (10 / 4) + 1) = 40 cost. 
+
+## Enchantment Effect Components
+
+Enchantment effect components are specially-registered [Data Components] that determine how an enchantment functions. The type of the component defines its effect, while the data it contains is used to inform or modify that effect. For instance, the `minecraft:damage` component modifies the damage that a weapon deals by an amount determined by its data.
+
+Vanilla defines various [built-in enchantment effect components], which are used to implement all vanilla enchantments.
+
+### Custom Enchantment Effect Components
+
+Enchantment effect component types must be [registered] to `BuiltInRegistries.ENCHANTMENT_EFFECT_COMPONENT_TYPE`, which takes a `DataComponentType<?>`. For example, you could register an enchantment effect component that can store an `ExampleEffectData` object as follows:
+
+```java
+// In some registration class
+public static final DeferredRegister<DataComponentType<?>> ENCHANTMENT_COMPONENT_TYPES = DeferredRegister.create(BuiltInRegistries.ENCHANTMENT_EFFECT_COMPONENT_TYPE, "examplemod");
+
+public static final DeferredHolder<DataComponentType<?>, DataComponentType<ExampleEffectData>>> EXAMPLE =
```

Comment:
And what is `ExampleEffectData`? This does not explain how a custom effect is made. Start with explaining that a custom enchantment effect component is implemented entirely by the user. The only thing that specifically needs to be registered is the data component that is stored within the enchantment itself. From there, the user must then search the specified object for enchantments and check if there enchantment is there. Then, implement the logic. Just creating the custom component does nothing.